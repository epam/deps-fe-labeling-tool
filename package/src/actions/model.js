
import { batch } from 'react-redux'
import { createAction } from 'redux-actions'
import { LabelType, LABEL_TYPE_NAME } from '@/models/Label'
import { Markup } from '@/models/Markup'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector, fieldsToDeleteSelector } from '@/selectors/model'
import {
  updateLabelsWithSettings,
  updateTablesWithSettings
} from './markup'

const FEATURE_NAME = 'MODEL'

const storeFields = createAction(
  `${FEATURE_NAME}/STORE`
)

const storeInitialFields = createAction(
  `${FEATURE_NAME}/STORE_INITIAL_FIELDS`
)

const updateInitialFields = createAction(
  `${FEATURE_NAME}/UPDATE_INITIAL_FIELDS`
)

const enableDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/ENABLE_DELETE_FIELD_MODE`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const updateFieldsToDelete = createAction(
  `${FEATURE_NAME}/UPDATE_FIELDS_TO_DELETE`
)

const cancelDeleteFieldsMode = createAction(
  `${FEATURE_NAME}/CANCEL_DELETE_FIELDS_MODE`
)

const confirmFieldsDeletion = () => (dispatch, getState) => {
  const state = getState()
  const markup = markupSelector(state)
  const fields = fieldsSelector(state)
  const fieldsToDelete = fieldsToDeleteSelector(state)
  const newFields = fields.filter((field) => fieldsToDelete.every((f) => f.code !== field.code))
  const assignedObjects = markup
    ? Markup.getAllObjects(markup).filter((markupObject) => fieldsToDelete.find((field) => field.code === markupObject.fieldCode))
    : []

  batch(() => {
    assignedObjects.forEach((markupObject) => {
      if (markupObject.typeName === LABEL_TYPE_NAME) {
        dispatch(updateLabelsWithSettings({
          ...markupObject,
          fieldCode: '',
          type: LabelType.UNASSIGNED,
          index: undefined
        }))
      } else {
        dispatch(updateTablesWithSettings({
          ...markupObject,
          fieldCode: '',
          index: undefined
        }))
      }
    })
    dispatch(storeFields(newFields))
    dispatch(cancelDeleteFieldsMode())
  })
}

export {
  enableDeleteFieldsMode,
  resetDefault,
  storeFields,
  storeInitialFields,
  updateFieldsToDelete,
  updateInitialFields,
  confirmFieldsDeletion,
  cancelDeleteFieldsMode
}
