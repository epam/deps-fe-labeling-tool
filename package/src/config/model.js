import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { PairFieldMeta, ListFieldMeta, EnumFieldMeta } from '@/models/FieldMeta'

const fields = [
  new Field(
    'dates',
    'Dates',
    FieldType.DATE
  ),
  new Field(
    'firstName',
    'First Name',
    FieldType.STRING
  ),
  new Field(
    'lastName',
    'Last Name',
    FieldType.STRING,
    undefined,
    true
  ),
  new Field(
    'companyName',
    'Company Name',
    FieldType.STRING,
    undefined,
    true
  ),
  new Field(
    'city',
    'City',
    FieldType.ENUM,
    new EnumFieldMeta([
      'Anytown',
      'Sometown',
      'San Diego',
      'Los Angeles'
    ]),
    true
  ),
  new Field(
    'countryStates',
    'Country States',
    FieldType.LIST,
    new ListFieldMeta(FieldType.STRING),
    true
  ),
  new Field(
    'dateMatching',
    'Date Matching',
    FieldType.LIST,
    new ListFieldMeta(FieldType.CHECKMARK),
    true
  ),
  new Field(
    'date',
    'Date',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING)
  ),
  new Field(
    'zipCodes',
    'Zip Codes',
    FieldType.LIST,
    new ListFieldMeta(
      FieldType.PAIR,
      new PairFieldMeta(FieldType.STRING, FieldType.STRING)
    ),
    true
  ),
  new Field(
    'homeAddress',
    'Home Address',
    FieldType.STRING
  ),
  new Field(
    'businessAddress',
    'Business Address',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING),
    true
  ),
  new Field(
    'taxpayerSpousesInfo',
    'Taxpayer Spouse\'s Info',
    FieldType.TABLE
  ),
  new Field(
    'taxpayerInfo',
    'Taxpayer Info',
    FieldType.TABLE,
    undefined,
    true
  ),
  new Field(
    'tables',
    'Tables',
    FieldType.LIST,
    new ListFieldMeta(FieldType.TABLE)
  ),
  new Field(
    'uniqueIdentifier',
    'Unique Identifier',
    FieldType.PAIR,
    new PairFieldMeta(FieldType.STRING, FieldType.STRING)
  ),
  new Field(
    'signatory',
    'Signatory',
    FieldType.CHECKMARK
  ),
  new Field(
    'returnTranscript',
    'Return Transcript',
    FieldType.CHECKMARK
  ),
  new Field(
    'formNumbers',
    'Form Numbers',
    FieldType.LIST,
    new ListFieldMeta(
      FieldType.ENUM,
      new EnumFieldMeta([
        '1098',
        '5498',
        '1040'
      ])
    )
  )
]

export {
  fields
}
