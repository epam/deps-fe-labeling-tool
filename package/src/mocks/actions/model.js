const storeFields = jest.fn((fields) => ({
  type: 'MOCK_STORE_FIELDS_ACTION',
  payload: fields
}))

const updateFieldsToDelete = jest.fn((fields) => ({
  type: 'MOCK_UPDATE_FIELDS_TO_DELETE_ACTION',
  payload: fields
}))

const storeInitialFields = jest.fn((fields) => ({
  type: 'MOCK_STORE_INITIAL_FIELDS_ACTION',
  payload: fields
}))

const mockModelActions = {
  storeFields,
  storeInitialFields,
  updateFieldsToDelete
}

export {
  mockModelActions
}
