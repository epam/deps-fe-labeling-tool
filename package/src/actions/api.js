import { createAction } from 'redux-actions'
import { rotationAnglesSelector } from '@/selectors/canvas'
import { pageImageUrlSelector } from '@/selectors/document'
import { markupSelector } from '@/selectors/markup'
import { fieldsSelector } from '@/selectors/model'
import { primaryLanguageSelector } from '@/selectors/ocr'
import { getApi } from '@/services/api'

const FEATURE_NAME = 'API'

const requestAttempt = createAction(
  `${FEATURE_NAME}/REQUEST_ATTEMPT`
)

const requestSuccess = createAction(
  `${FEATURE_NAME}/REQUEST_SUCCESS`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const requestFailure = createAction(
  `${FEATURE_NAME}/REQUEST_FAILURE`,
  (requestId, error) => ({
    requestId,
    error
  })
)

export const createRequestAction = (requestId, actionCreator) => {
  const requestActionCreator = (...args) => async (dispatch, getState) => {
    try {
      dispatch(requestAttempt(requestId))
      const result = await actionCreator(...args)(dispatch, getState)
      dispatch(requestSuccess(requestId))
      return result
    } catch (error) {
      dispatch(requestFailure(requestId, error.message))
      throw error
    }
  }
  requestActionCreator.toString = () => requestId
  return requestActionCreator
}

const save = createRequestAction(
  'save',
  () => (dispatch, getState) => {
    const state = getState()
    const markup = markupSelector(state)
    const rotationAngles = rotationAnglesSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().save(markup, rotationAngles, language)
  }
)

const saveMarkup = createRequestAction(
  'saveMarkup',
  () => (dispatch, getState) => {
    const state = getState()
    const markup = markupSelector(state)
    const rotationAngles = rotationAnglesSelector(state)
    const language = primaryLanguageSelector(state)
    const fields = fieldsSelector(state)
    return getApi().saveMarkup(markup, rotationAngles, language, fields)
  }
)

const ocrText = createRequestAction(
  'ocrText',
  (engine, coordinates) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().ocrText(
      engine,
      pageImageUrl,
      coordinates,
      language
    )
  }
)

const omrArea = createRequestAction(
  'omrArea',
  (coordinates) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)

    return getApi().omrArea(
      pageImageUrl,
      coordinates
    )
  }
)

const detectTables = createRequestAction(
  'detectTables',
  (selectedAreaCoords) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    return getApi().detectTables(
      pageImageUrl,
      selectedAreaCoords
    )
  }
)

const ocrTable = createRequestAction(
  'ocrTable',
  (engine, table) => (dispatch, getState) => {
    const state = getState()
    const pageImageUrl = pageImageUrlSelector(state)
    const language = primaryLanguageSelector(state)
    return getApi().ocrTable(
      engine,
      pageImageUrl,
      table,
      language
    )
  }
)

export {
  detectTables,
  ocrTable,
  ocrText,
  omrArea,
  resetDefault,
  requestAttempt,
  requestSuccess,
  requestFailure,
  save,
  saveMarkup
}
