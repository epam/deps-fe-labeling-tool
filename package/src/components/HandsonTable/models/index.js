
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
