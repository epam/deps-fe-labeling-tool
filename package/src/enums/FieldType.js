const FieldType = {
  CHECKMARK: 'checkmark',
  ENUM: 'enum',
  LIST: 'list',
  PAIR: 'pair',
  STRING: 'string',
  TABLE: 'table',
  DATE: 'date'
}

const RESOURCE_FIELD_TYPE = {
  [FieldType.STRING]: 'String',
  [FieldType.ENUM]: 'Enumeration',
  [FieldType.CHECKMARK]: 'Checkbox',
  [FieldType.LIST]: 'List',
  [FieldType.PAIR]: 'Key Value Pair',
  [FieldType.TABLE]: 'Table',
  [FieldType.DATE]: 'Date'
}

export {
  FieldType,
  RESOURCE_FIELD_TYPE
}
