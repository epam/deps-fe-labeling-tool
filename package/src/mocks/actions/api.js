import { mockMarkupSelectors } from '@/mocks/selectors/markup'

const omrArea = jest.fn(() => Promise.resolve({
  content: false,
  confidence: 0.77
}))

const ocrTable = jest.fn(() => Promise.resolve({ type: 'MOCK_OCR_TABLE' }))

const ocrText = jest.fn(() => Promise.resolve({
  content: 'MOCK_OCR_TEXT',
  confidence: 0.98
}))

const save = jest.fn(() => Promise.resolve(mockMarkupSelectors.markupSelector()))

const saveMarkup = jest.fn(() => Promise.resolve({ type: 'MOCK_SAVE_MARKUP' }))

const detectTables = jest.fn((url, coords) => Promise.resolve(['markupTable']))

const mockApiActions = {
  detectTables,
  omrArea,
  ocrTable,
  ocrText,
  save,
  saveMarkup
}

export {
  mockApiActions
}
