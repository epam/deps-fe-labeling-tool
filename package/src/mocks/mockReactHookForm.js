import { FieldType } from '@/enums/FieldType'

const mockReactHookForm = ({
  ...jest.requireActual('react-hook-form'),
  useForm: jest.fn(() => ({
    handleSubmit: jest.fn(),
    getValues: jest.fn(),
    formState: {}
  })),
  useController: jest.fn(() => ({
    field: {},
    fieldState: {}
  })),
  useFormContext: jest.fn(() => ({
    control: {},
    setValue: jest.fn()
  })),
  useWatch: jest.fn(() => FieldType.STRING)
})

export {
  mockReactHookForm
}
