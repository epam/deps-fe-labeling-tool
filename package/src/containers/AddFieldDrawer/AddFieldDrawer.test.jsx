import { mockModelActions } from '@/mocks/actions/model'
import { mockReactHookForm } from '@/mocks/mockReactHookForm'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockModelSelectors } from '@/mocks/selectors/model'
import React from 'react'
import { shallow } from 'enzyme'
import { useForm } from 'react-hook-form'
import { storeFields } from '@/actions/model'
import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { EnumFieldMeta, ListFieldMeta } from '@/models/FieldMeta'
import { getApi } from '@/services/api'
import {
  AddButton,
  AddFieldButton,
  Drawer
} from './AddFieldDrawer.styles'
import { AddFieldDrawer } from '.'

const AddFieldForm = () => <div></div>
const mockNotify = { error: jest.fn() }

jest.mock('react-hook-form', () => mockReactHookForm)
jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/selectors/model', () => mockModelSelectors)
jest.mock('@/actions/model', () => mockModelActions)
jest.mock('@/services/api', () => ({
  getApi: jest.fn(() => ({
    addFieldForm: AddFieldForm,
    notify: mockNotify
  }))
}))

describe('Container: AddFieldDrawer', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<AddFieldDrawer />)
  })

  it('should render layout correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should call storeFields with correct argument in case creating a new field', () => {
    getApi.mockImplementationOnce(() => ({}))

    const mockedResult = {
      fieldMeta: new ListFieldMeta(FieldType.TABLE),
      code: 'code',
      fieldType: FieldType.TABLE,
      name: 'name',
      required: 'test'
    }

    const mockFormValues = {
      code: 'code',
      name: 'name',
      fieldType: FieldType.TABLE,
      required: 'test'
    }

    useForm.mockImplementation(() => ({
      formState: {},
      getValues: jest.fn(() => mockFormValues)
    }))

    wrapper = shallow(<AddFieldDrawer />)

    const Footer = wrapper.find(Drawer).props().footer
    const DrawerFooter = shallow(<div>{Footer}</div>)
    DrawerFooter.find(AddButton).props().onClick()

    expect(storeFields).nthCalledWith(
      1,
      [
        ...mockModelSelectors.fieldsSelector(),
        mockedResult
      ]
    )
  })

  it('should call storeFields with correct argument in case creating a new field with enum type', () => {
    getApi.mockImplementationOnce(() => ({}))
    const mockOptions = ['option']

    const mockedResult = {
      fieldMeta: new EnumFieldMeta(mockOptions),
      code: 'code',
      fieldType: FieldType.ENUM,
      name: 'name',
      required: 'test'
    }

    const mockFormValues = {
      code: 'code',
      name: 'name',
      fieldType: FieldType.ENUM,
      required: 'test',
      options: mockOptions
    }

    useForm.mockImplementation(() => ({
      formState: {},
      getValues: jest.fn(() => mockFormValues)
    }))

    wrapper = shallow(<AddFieldDrawer />)

    const Footer = wrapper.find(Drawer).props().footer
    const DrawerFooter = shallow(<div>{Footer}</div>)
    DrawerFooter.find(AddButton).props().onClick()

    expect(storeFields).nthCalledWith(
      1,
      [
        ...mockModelSelectors.fieldsSelector(),
        mockedResult
      ]
    )
  })

  it('should change Drawer visibility when click on onToggleVisible', () => {
    getApi
      .mockImplementationOnce(() => ({}))
      .mockImplementationOnce(() => ({}))

    wrapper = shallow(<AddFieldDrawer />)

    const previousVisibility = wrapper.find(Drawer).props().visible

    wrapper.find(AddFieldButton).props().onClick()

    expect(wrapper.find(Drawer).props().visible).not.toBe(previousVisibility)
  })

  it('should notify User of error if field already exists when running onSave', () => {
    const addFieldForm = wrapper.find(getApi().addFieldForm).props()
    const field = mockModelSelectors.fieldsSelector()[0]
    const ERROR_MESSAGE = 'Field already exists'
    addFieldForm.onSave(field)
    expect(getApi().notify?.error).nthCalledWith(1, ERROR_MESSAGE)
  })

  it('should correctly store new field when running onSave', () => {
    const addFieldForm = wrapper.find(getApi().addFieldForm).props()
    const fields = mockModelSelectors.fieldsSelector()
    const field = new Field('1', '1', FieldType.STRING)
    addFieldForm.onSave(field)
    expect(storeFields).toHaveBeenNthCalledWith(1, [...fields, field])
  })

  it('should render layout correctly if getApi().addFieldForm is undefined', () => {
    getApi.mockImplementationOnce(() => ({}))

    wrapper = shallow(<AddFieldDrawer />)

    expect(wrapper.find(Drawer).exists()).toBe(true)
  })
})
