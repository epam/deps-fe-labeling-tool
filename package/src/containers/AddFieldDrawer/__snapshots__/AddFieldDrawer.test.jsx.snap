// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: AddFieldDrawer should render layout correctly 1`] = `
<AddFieldDrawerstyles__AddFieldFormWrapper>
  <AddFieldForm
    onSave={[Function]}
  >
    <PlusIcon />
    <AddFieldDrawerstyles__AddFieldFormContext>
      Add field
    </AddFieldDrawerstyles__AddFieldFormContext>
  </AddFieldForm>
</AddFieldDrawerstyles__AddFieldFormWrapper>
`;
